Add CML m/c for RT1011 and RT5682

sathyap-chrome opened the pull request:
This machine driver uses RT5682 on SSP0 for HS / AMIC and RT1011 on SSP1.
RT1011 is smart amp which drives 2 CH data over 4 individual speakers.

plbossart commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,519 @@
+struct cml_card_private {
+	struct cml_acpi_card *acpi_card;
+	char codec_name[SND_ACPI_I2C_ID_LEN];
+	struct clk *mclk;
```
where is this mclk configured?

sathyap-chrome replied:
agreed, my bad. mclk config is from topology and hence no clk configuration is needed here.

plbossart commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,519 @@
+	if (!ret)
+		clk_disable_unprepare(ctx->mclk);
+
+	ret = clk_set_rate(ctx->mclk, CML_PLAT_CLK);
```
we don't have an implementation to turn on the MCLK on the Intel DSP side?
What is this code supposed to do here?

sathyap-chrome replied:
fixed in next version.

plbossart commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,519 @@
+	},
+	{
+		.name = "i2c-10EC1011:03",
+		.dai_name = "rt1011-aif",
```
can you clarify what this is supposed to do?
Do we have 4 amplifiers connected to SSP1?

sathyap-chrome replied:
yes. 4 AMPS are connected to SSP1

plbossart commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,519 @@
+		.num_codecs = ARRAY_SIZE(rt1011_codec_component),
+		.init = cml_rt1011_codec_init,
+		.dpcm_playback = 1,
+		.dpcm_capture = 1,
```
what is the capture used for? Echo reference/IV sensing? If yes, please add a comment here.

sathyap-chrome replied:
Yes, it is for Echo reference . I'll add the comment

plbossart commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,519 @@
+			found = true;
+			break;
+		}
+	}
```
this is not really needed for now, you have a single card to manage...

sathyap-chrome replied:
fixed in next version

plbossart commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,519 @@
+	struct cml_card_private *ctx;
+	struct snd_soc_acpi_mach *mach;
+	const char *platform_name;
+	struct snd_soc_card *card = snd_soc_cards[0].soc_card;
```
is the init needed?

plbossart commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,519 @@
+		}
+
+		/* TDM 4 slots 24 bit, set Rx & Tx bitmask to 4 active slots */
+		ret = snd_soc_dai_set_tdm_slot(codec_dai, 0xF, 0xF, 4, 24);
```
does this make sense to configure all 4 codecs to use 4 slots? You'd want each codec to use one out of 4?

sathyap-chrome replied:
The codec config part i refereed the vendor implementation. 
According to me, I believe all codec will have same configuration wrt no of slots to drive the necessary clks. Codec driver exposes Kcontrols to pick right slot for right AMP . Maybe below reference will help 
https://git.kernel.org/pub/scm/linux/kernel/git/broonie/sound.git/tree/sound/soc/codecs/rt1011.c?h=for-next

And below UCM controls the slots for each AMP
        # TL/TR/WL/WR default settings
		cset "name='TL TDM1 DOUT Location' 0"
		cset "name='TR TDM1 DOUT Location' 1"
		cset "name='WL TDM1 DOUT Location' 2"
		cset "name='WR TDM1 DOUT Location' 3"

let me know if my understanding is not right.

plbossart replied:
I meant something like ret = snd_soc_dai_set_tdm_slot(codec_dai, BIT(i), BIT(i), 4, 24);

Why would you tell all codecs that they should look at all the slots?

ranj063 commented:
@sathyap-chrome please also fix checkpatch issues

ranj063 commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,519 @@
+
+	/* DMIC */
+	{"DMic", NULL, "SoC DMIC"},
+
```
@sathyap-chrome no need for a new line

sathyap-chrome replied:
Done

ranj063 commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,519 @@
+		dev_err(rtd->dev, "snd_soc_dai_set_sysclk err = %d\n", ret);
+
+	/*
+	 * slot_width should equal or large than data length, set them
```
typo here..."slot_width should be"

sathyap-chrome replied:
done

ranj063 commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,519 @@
+		return ret;
+	}
+
+	return ret;
```
@sathyap-chrome no need for 2 return statement. just "return ret" outside the if

sathyap-chrome replied:
done

ranj063 replied:
@sathyap-chrome hmm I still see 2 return statements? What I meant was: 

if (ret < 0)
dev_err(rtd->dev, "set TDM slot err:%d\n", ret);

return ret;

ranj063 commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,519 @@
+		.cpu_dai_name = "iDisp3 Pin",
+		.codec_name = "ehdaudio0D2",
+		.codec_dai_name = "intel-hdmi-hifi3",
+		.platform_name = "0000:00:1f.3",
```
@sathyap-chrome the machine drivers have been modified to use modern dai link representation. So this needs to be changed here as well. Please refer to bxt_da7219_max98357a.c for examples of how to do this.

sathyap-chrome replied:
Thanks for reference. this is updated in new version

ranj063 commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,519 @@
+
+static struct platform_driver snd_cml_rt1011_rt5682_driver = {
+	.driver = {
+		.name = "cml_rt1011_rt5682",
```
@sathyap-chrome 		add missing ".pm = &snd_soc_pm_ops,"

sathyap-chrome replied:
my bad. added

ranj063 commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,519 @@
+	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct snd_soc_dai *codec_dai = rtd->codec_dai;
+	struct snd_soc_card *card = rtd->card;
+	int err, srate, i, ret;
```
@sathyap-chrome no need for err and ret. Can use just one or the other.

sathyap-chrome replied:
done

ranj063 commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,519 @@
+	struct snd_soc_component *component = NULL;
+	char jack_name[NAME_SIZE];
+	struct cml_hdmi_pcm *pcm;
+	int err = 0;
```
@sathyap-chrome no need to initialize err

sathyap-chrome replied:
done

ranj063 commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,519 @@
+		return ret_val;
+	}
+
+	platform_set_drvdata(pdev, card);
```
@sathyap-chrome not needed. Just "return devm_snd_soc_register_card(&pdev->dev, &snd_soc_card_cml);" here

sathyap-chrome replied:
Sure @ranj063  will address all comments from you and Pierre and repost.

ranj063 commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,449 @@
+#include <sound/soc.h>
+#include <sound/rt5682.h>
+#include <sound/soc-acpi.h>
+#include "../skylake/skl.h"
```
@sathyap-chrome do you really need this?

sathyap-chrome replied:
will fix it

ranj063 commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,449 @@
+		SND_SOC_DAILINK_REG(idisp3_pin, idisp3_codec, platform),
+	},
+	{
+		/* SSP1 - Codec */
```
@sathyap-chrome this ordering of dai links seems odd. Any particular reason why you dont keep the 2 SSP's together?

sathyap-chrome replied:
i can reorder, but the topology uses similar ordering and hence maintained that.

plbossart replied:
please reorder and fix the topology. Things are only required to be stable when both are approved

sathyap-chrome replied:
sure

ranj063 commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,449 @@
+	snd_soc_card_set_drvdata(&snd_soc_card_cml, ctx);
+
+	return devm_snd_soc_register_card(&pdev->dev, &snd_soc_card_cml);
+
```
@sathyap-chrome unnecessary newline

sathyap-chrome replied:
will fix it

plbossart commented on sound/soc/intel/boards/Kconfig:
```diff
@@ -437,6 +437,21 @@ config SND_SOC_INTEL_SOF_RT5682_MACH
+	tristate "CML SOF with RT1011 in I2S Mode"
+	depends on I2C && ACPI
+	depends on (SND_SOC_SOF_HDA_COMMON && MFD_INTEL_LPSS) ||\
+	           (SND_SOC_SOF_BAYTRAIL && X86_INTEL_LPSS)
```
remove baytrail

plbossart replied:
align with other machine drivers:

depends on I2C && ACPI
depends on MFD_INTEL_LPSS || COMPILE_TEST

plbossart replied:
and put this outside of the if statement with BAYTRAIL.

GAH.

sathyap-chrome replied:
my bad on the config placement. will correct

plbossart commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,449 @@
+ * Intel Cometlake I2S Machine driver for RT1011 + RT5682 codec
+ *
+ * Modified from:
+ *   Intel I2S Machine driver for RT5682 + Maxim98357a
```
remove this. it's useless

sathyap-chrome replied:
sure. will do

plbossart commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,449 @@
+
+static struct snd_soc_jack sof_hdmi[3];
+
+struct cml_hdmi_pcm {
```
remove all those cml_ prefix, they are useless except for controls, widgets and card names

sathyap-chrome replied:
sure. will fix

sathyap-chrome commented:
Addressed all review comments.
 Also for TDM changes, corresponding codec driver changes were posted here:
https://patchwork.kernel.org/patch/11148169/

plbossart commented on sound/soc/intel/boards/Kconfig:
```diff
@@ -437,6 +437,7 @@ config SND_SOC_INTEL_SOF_RT5682_MACH
 	   with rt5682 codec.
 	   Say Y if you have such a device.
 	   If unsure select "N".
+
```
spurious newline, please remove

plbossart commented on sound/soc/intel/boards/Kconfig:
```diff
@@ -466,4 +467,17 @@ config SND_SOC_INTEL_TGL_RT1308_MACH
 
+config SND_SOC_INTEL_SOF_CML_RT1011_RT5682_MACH
+        tristate "CML with RT1011 and RT5682 in I2S Mode"
+        depends on MFD_INTEL_LPSS && I2C && ACPI
```
I've already mentioned you need to align with other machine drivers and add COMPILE_TEST.

plbossart commented on sound/soc/intel/boards/Kconfig:
```diff
@@ -466,4 +467,17 @@ config SND_SOC_INTEL_TGL_RT1308_MACH
+        select SND_SOC_RT1011
+        select SND_SOC_RT5682
+        select SND_SOC_DMIC
+        select SND_SOC_HDAC_HDMI if SND_SOC_SOF_HDA_COMMON
```
if SND_SOC_SOF_HDA_LINK

plbossart commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,477 @@
+#define CML_RT5682_CODEC_DAI "rt5682-aif1"
+#define NAME_SIZE 32
+
+static struct snd_soc_jack sof_hdmi[3];
```
remove the sof_ prefix.

plbossart commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,477 @@
+		}
+		if (!strcmp(codec_dai->component->name, "i2c-10EC1011:03")) {
+			ret = snd_soc_dai_set_tdm_slot(codec_dai,
+						       0x2, 0x2, 4, 24);
```
that RX slot is the same as the :01 device, so this cannot possibly work.
Why can't you use a more regular loop and slot assignments instead of those comparisons on the name and different slots for RX and TX?

sathyap-chrome replied:
The RX slot for 2 codecs will be same as its a 2 CH playback over 4 codecs. But each codec provides its own feedback. So each codec Tx is unique, but 2 codecs will have same Rx.

sathyap-chrome replied:
Just to add more details, the 2 CH playback is done over 4 codecs - 2 Woofers and 2 Tweeters. 
The Left woofer and tweeter plays the Left playback data and similar by the Right. The RT1011 internally manages control of the Tweeter & Woofer and parameters for the same.

Hence 2 codecs (1 T and 1 W pair) share same Rx slot. I've updated the title too explaining we  are playing 2 CH TDM. Feedback is 4 CH.

Let me know if any more details needed.

plbossart replied:
in that case, please make it explicit which codecs use what slot for RX, e.g. left RX first and then right RX second. If reviewers can't figure out a feature and think you've done a mistake there's something missing.

sathyap-chrome replied:
Sure Pierre. I've updated with explaining the rationale and grouping based on Rx slot as suggested.
I

plbossart commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,477 @@
+		.dpcm_playback = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(idisp3_pin, idisp3_codec, platform),
+	},
```
if you select HDAC_HDMI conditionally in the Kconfig, you need to conditionally add HDMI here as well.

ranj063 replied:
@sathyap-chrome Pierre's comment applies to a couple of other places as well. like the sof_card_late_probe(), line 447 etc. Please take a look at tgl_rt1308.c for example.

sathyap-chrome replied:
I corrected the Kconfig aligning to other machine driver so the condition is not required. As platform is having HDMI.

ranj063 commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,477 @@
+				       params_width(params));
+	if (ret < 0) {
+		dev_err(rtd->dev, "set TDM slot err:%d\n", ret);
+	}
```
@sathyap-chrome remove braces here. checkpatch must have caught this isnt it?

sathyap-chrome replied:
sure i will. but checkpatch never caught this

sathyap-chrome replied:
fixed

ranj063 commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,477 @@
+		ret = snd_soc_card_jack_new(card, jack_name,
+					    SND_JACK_AVOUT, &sof_hdmi[i],
+					    NULL, 0);
+
```
remove newline

sathyap-chrome replied:
Done

wenqingfu commented:
SOFCI TEST

plbossart commented on sound/soc/intel/boards/Kconfig:
```diff
@@ -466,4 +466,18 @@ config SND_SOC_INTEL_TGL_RT1308_MACH
+        select SND_SOC_RT1011
+        select SND_SOC_RT5682
+        select SND_SOC_DMIC
+        select SND_SOC_HDAC_HDMI
```
take 3: please make this configuration dependent on SOF_HDA_LINK if you unconditionally select HDMI.

plbossart commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,475 @@
+			if (ret < 0) {
+				dev_err(rtd->dev,
+					"can't set codec TDM slot %d\n", ret);
+				return ret;
```
as mentioned above, please group by RX slots and add comments so that the use of slots is self-explanatory.

sathyap-chrome replied:
sure pierre. will do that

plbossart commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,475 @@
+		.no_pcm = 1,
+		.ignore_suspend = 1,
+		.dpcm_capture = 1,
+		.no_pcm = 1,
```
duplicate .no_pcm

plbossart commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,485 @@
+	ret = snd_soc_dai_set_sysclk(codec_dai, RT5682_SCLK_S_PLL1,
+				     pll_out, SND_SOC_CLOCK_IN);
+	if (ret < 0)
+		dev_err(rtd->dev, "snd_soc_dai_set_sysclk err = %d\n", ret);
```
you don't return when setting PLL or sysclck (lines 137 and 143), is this intentional?
if yes, this should be dev_warn(), otherwise you need to add a return ret;

sathyap-chrome replied:
Sure Pierre. i can do that. i used the sof-rt5682.c as reference to program the 5682 portion here.

plbossart replied:
let's hear from the initial author. @bardliao can you comment, is this intentional or should we change sof_rt5682 as well?

bardliao replied:
@plbossart @sathyap-chrome In my opinion, dev_warn() is good enough since rt5682 can still play (maybe with noise) even clock configuration is wrong. And people can know there are something wrong from the trace.

sathyap-chrome replied:
Thanks @bardliao .  @plbossart with latest push i've changed to dev_warn().

ranj063 commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,485 @@
+		 * The Left woofer and tweeter plays the Left playback data
+		 * and  similar by the Right.
+		 * Hence 2 codecs (1 T and 1 W pair) share same Rx slot.
+		 * The feedback is captured for each codec individaully.
```
@typo here "individually"

sathyap-chrome replied:
thanks for pointing .. will correct :)

ranj063 commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,485 @@
+		 * Hence all 4 codecs use 1 Tx slot each for Feedback.
+		 */
+
+		/* TDM Rx slot 1 is used for Left Woofer & Tweeters pair */
```
combine the 2 comments into the same block. Dont seem to be picky but please dont use words starting with capital letters in the middle of a sentence. like "Hence, all 4 codecs use 1 Tx slot each for Feedback"

plbossart replied:
It's actually not so bad to use a capital letter when the word refers to an object or concept that has a specific definition.
E.g. if there is a 'Feedback' stream then it's ok to use the capital. If the feedback stream is called 'IV measurements' then it's not.
Standards typically use this convention, e.g. in the SoundWire spec when you see 'Slave' with a capital letter its definition is unambiguous and described in the glossary.

ranj063 replied:
@plbossart absolutely but in the example above there isnt a need for capitalizing "Feedback". Wherewas in the following sentence, capitalizing Left/Right is actually a good thing to do.

sathyap-chrome replied:
@ranj063 based on above conversation, i can add feedback without using F. 
But i would still like to maintain the 2 comments separate as the 1st is explaining the logic in more detail and the 2nd is a 1 liner of following code and it has an equivalent when we program the Right.

ranj063 commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,485 @@
+						       0x2, 0x2, 4, 24);
+			if (ret < 0) {
+				dev_err(rtd->dev,
+					"can't set codec TDM slot %d\n", ret);
```
@sathyap-chrome this looks good. Just a minor comment. There are 4 errors with the same error mesage "can't set codec TDM slot %d". If one of them does fail, it would be hard to tell where the failure occurred. Can you add the component name maybe?

sathyap-chrome replied:
sure. will do that.

ranj063 commented on sound/soc/intel/boards/cml_rt1011_rt5682.c:
```diff
@@ -0,0 +1,485 @@
+				struct snd_pcm_hw_params *params)
+{
+	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct snd_soc_dai *codec_dai = rtd->codec_dai;
```
this initialization is also not required.

sathyap-chrome replied:
Done

ranj063 commented:
@sathyap-chrome can you please take care of checkpatch issues? click on the "Details" link in the CI errors

sathyap-chrome commented:
@ranj063  " I hope this has been actually tested on HW" - yes., Helios Proto and EVT. Testing every change made.

sathyap-chrome commented:
updated with the 1 checkpatch fix.

plbossart commented:
@xiulipan can you try to improve the CI status, it shows a non-working Up2-nocodec but the test was not run at all...
